Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp

```diff
@@ -28,49 +28,369 @@
+    
+    if (!start || !end || !contains(searchRange, start.value()) || !contains(searchRange, end.value()))
+        return std::nullopt;
     
-    auto resultRange = collapseIfRootsDiffer(findPlainText(searchRange, target, findOptions));
+    return SimpleRange(*start, *end);
+    
+}
+
+static std::optional<SimpleRange> rangeOfStringInRange(const String& query, SimpleRange& searchRange, WordBounded wordStartBounded, WordBounded wordEndBounded)
+{
+    while (!searchRange.collapsed()) {
+        Node* currentNode = &searchRange.startContainer();
+        
+        if (isNonSearchableSubtree(*currentNode)) {
+            if (auto newStart = nextSkippingChildrenInComposedTreeIgnoringUserAgentShadow(*currentNode))
+                searchRange.start = { *newStart, 0 };
+            else
+                return std::nullopt;
+            continue;
+        }
+        
+        if (!isVisibleTextNode(*currentNode)) {
+            do {
+                if (auto newStart = nextInComposedTreeIgnoringUserAgentShadow(*currentNode)) {
+                    searchRange.start = { *newStart, 0 };
+                    currentNode = newStart;
+                } else
+                    return std::nullopt;
+            } while (is<DocumentType>(currentNode));
+            continue;
+        }
+        
+        auto& blockAncestor = nearestBlockAncestor(*currentNode);
+        Vector<RefPtr<Node>> textNodeList;
+        while (currentNode->isDescendantOf(blockAncestor) && is_lteq(treeOrder(BoundaryPoint(*currentNode, 0), searchRange.end))) {
+            if (currentNode->renderer() && is<Element>(currentNode) && currentNode->renderer()->style().isDisplayBlockLevel())
+                break;
 
-    return resultRange.collapsed() ? std::nullopt : std::make_optional(resultRange);
+            if (isSearchInvisible(*currentNode)) {
+                if (auto newNode = nextInComposedTreeIgnoringUserAgentShadow(*currentNode))
+                    currentNode = newNode;
+                else
+                    break;
+                continue;
+            }
+            if (currentNode->isTextNode())
+                textNodeList.append(currentNode);
+            if (auto newNode = nextInComposedTreeIgnoringUserAgentShadow(*currentNode))
+                currentNode = newNode;
+            else
+                break;
+        }
+
+        if (auto range = findRangeFromNodeList(query, searchRange, textNodeList, wordStartBounded, wordEndBounded))
+            return range;
+
+        if (!currentNode)
+            break;
+        
+        ASSERT_WITH_MESSAGE(is_gt(treeOrder(*currentNode, searchRange.start.container)), "Scroll To Text Fragment: Current node must follow start.");
+        
+        if (!is_gt(treeOrder(*currentNode, searchRange.start.container)))
+            return std::nullopt;
+        
+        searchRange.start = { *currentNode, 0 };
+    }
+    return std::nullopt;
+}
+
+static void advanceRangeStartToNextNonWhitespace(SimpleRange& range)
```

Comment:
Drop a link to https://wicg.github.io/scroll-to-text-fragment/#next-non-whitespace-position?